Schools Create and Index Page and Fixed Backend Issues

SR33G33 opened the pull request:
Closes #30 
Closes #29 

Implemented the path for Admin to go to school management page. The NavBar takes the admin to the schools index page

![image](https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4/assets/51520358/2bac1615-0ceb-4f7a-8292-aa3aae19d146)


Create School button leads to the create page for schools which works and successfully adds the school to the index page and the db

![image](https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4/assets/51520358/8b432114-94f4-43fe-867a-5903215ea282)

Back to index page

![image](https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4/assets/51520358/32649846-689d-4c80-b2f1-7a1554a88bd7)



All course staff and group members should have admin access



dev deployment at this link: https://proj-organic-sr33g33-dev.dokku-12.cs.ucsb.edu/




In the backend, I changed the schoolController to take in a school json object instead of @getParams because of an issue I was facing in testing where the received input was undefined

SR33G33 committed:
ss - created the create page for school and linked to the app.js file, having an issue with params not sending to backend

SR33G33 committed:
ss - removed .env.SAMPLE

SR33G33 committed:
ss - fixed params in create page

SR33G33 committed:
ss - removed commented code for edit page to be used in next PR

SR33G33 committed:
removed .env sample

SR33G33 committed:
ss - added test file for SchoolCreatePage

SR33G33 committed:
fixed backend issues that were causing Create Tests to fail

SR33G33 committed:
ss - fixed some SchoolController backend stuff for CreateTests to pass

SR33G33 committed:
ss - fixed syntax error in SchoolController

SR33G33 committed:
ss - accidentally changed invalidRegex test to ucsb instead of UCSB

SR33G33 committed:
ss - attempted fix for SchoolControllerTests so it accepts the new JSON format for the school

SR33G33 committed:
ss - more fixes to SchoolControllerTests, forgot to change base admin put test and added termRegex invalidity check

SR33G33 committed:
ss - added requestBodyCheck to admin can create school test

jason-rodrigues reviewed (commented):
I think you may have mixed up the description and regex checkers.
Could someone else confirm/deny this?

SR33G33 commented:
I believe the term description needs to follow the regex you inputted, so for your inputted regex the term description would simply be S24, F23 

![image](https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4/assets/51520358/0c584b13-15ed-401d-ac6c-baa4b8408777)

Like so

jason-rodrigues reviewed (dismissed):
LGTM

SR33G33 committed:
ss - accidentally deleted in previous commit, resetting this file from main branch

SR33G33 committed:
ss - Update CoursesEditPage.stories.js to be same as one in main branch

SR33G33 committed:
ss - Update CoursesEditPage.stories.js

sanilkatula commented:
<img width="785" alt="Screenshot 2024-05-30 at 5 29 01 PM" src="https://github.com/ucsb-cs156-s24/proj-organic-s24-5pm-4/assets/124859948/099b45a6-4a5f-4c03-b023-ce17f424b9ef">


This should not be happening.

Term Regex and Term Description cant be this. 

so the problem is when you are actually creating the school everything works fine and regex is checked properly but when you update the school its not checking for regex properly. Maybe try adding regex checks on the edit page as well.

sanilkatula reviewed (changes requested):
check CR

SR33G33 commented:
I added this as a bug which I pointed out in my edit page PR, I added it as a new task on the kanban board and am fixing it in a separate PR, I believe that the index and create pages are correct

pconrad commented:
The create school button leads to an error message for me on the dokku deployment linked in the PR.

I can't merge this unless/until I can test it.

SR33G33 commented:
redeployed on dokku seems to be working on my end

pconrad reviewed (changes requested):
You cannot delete .env.SAMPLE.  It is an essential file for the app.

Also: is it the intention in this PR to have the index page and create feature for school but to have edit and delete both return errors?

That's typically not mergeable.  We ask that the button at least link to a placeholder page, or put up a message sayig "not yet implemented".

I might relent in this case, but it's bad practice.  So I need clarity on that.

SR33G33 committed:
ss - accidentally deleted in old commit, resetting this file from main branch

SR33G33 committed:
Update .env.SAMPLE

SR33G33 commented:
I accidentally deleted the .env.SAMPLE file, I restored it from the version that is on the main branch

The dokku deployment now has the branch version that has the real functionality for edit and delete instead of errors. 

I should have inserted placeholder pages however I originally intended to have implemented all 3 functionalities in the same PR but later decided to break up the PR's for organizations sake so I just never got around to putting in placeholders as I just did them one after another.

pconrad commented on src/main/java/edu/ucsb/cs156/organic/controllers/SchoolController.java:
```diff
@@ -112,34 +111,21 @@ public Object deleteSchool(
             throw new IllegalArgumentException("Invalid abbrev format. Abbrev must be all lowercase");
         }
 
+        if (!school.getTermDescription().matches(school.getTermRegex())) {
+            throw new IllegalArgumentException("Invalid termDescription format. It must follow the pattern " + school.getTermRegex());
+        }
+
```
As long as we are editing this file, let's fix this, because this was never the intention of this field.

The term description is just supposed to be the word "Quarter" (e.g. for UCSB) or "Semester" (e.g. for Washington State).

Just remove this.

```suggestion
```

pconrad commented on frontend/src/tests/pages/SchoolCreatePage.test.js:
```diff
@@ -0,0 +1,114 @@
+        fireEvent.change(abbrevField, { target: { value: 'ucsb' } });
+        fireEvent.change(nameField, { target: { value: 'UC Santa Barbara' } });
+        fireEvent.change(termRegexField, { target: { value: '[WSMF]\\d\\d' } });
+        fireEvent.change(termDescriptionField, { target: { value: 's24' } });
```
```suggestion
        fireEvent.change(termDescriptionField, { target: { value: 'quarter' } });
```

pconrad commented on frontend/src/tests/pages/SchoolCreatePage.test.js:
```diff
@@ -0,0 +1,114 @@
+                "abbrev": "ucsb",
+                "name": "UC Santa Barbara",
+                "termRegex": "[WSMF]\\d\\d",
+                "termDescription": "s24",
```
```suggestion
                "termDescription": "quarter",
```

pconrad reviewed (changes requested):
Looks much better! There is still one issue that we should address: a misunderstanding about the use case / purpose for these fielods.

SR33G33 committed:
ss - TermDescription is a field that just determines term type, has nothing to do with term regex

SR33G33 committed:
ss - changed backend tests to match this update

SR33G33 committed:
ss - should fix mutation testing for backend

SR33G33 committed:
ss - attempted fix for backend test mutation failing

SR33G33 committed:
ss - more minor backend test fixes for clarity

SR33G33 committed:
ss - another attempt to fix backend mutation test

SR33G33 commented:
Fixed it, now TermDescription has nothing to do with the TermRegex and the corresponding tests and checks have been modified/removed, will update the later PRs with this as well

SR33G33 committed:
ss - the pitfalls of blindly find and replacing, minor fixes to backend tests for clarity

SR33G33 committed:
ss - another attempted fix for mutation

pconrad reviewed (approved):
LGTM
